### PR TITLE
Add UPnP and NAT-PMP support

### PR DESCRIPTION
**Pull request by Ansa89**
This supersedes irungentoo/toxcore#1537 and is partially a rebase of irungentoo/toxcore#969.

### CODE REVIEW

**Comment by Ansa89** [comment]
`TOX_TRAVERSAL_TYPE` should be defined directly into [tox.h](toxcore/tox.h).
This is a bad error, I will try to fix it ASAP.

**Comment by Ansa89** [comment]
> `TOX_TRAVERSAL_TYPE` should be defined directly into `toxcore/tox.h`.
> This is a bad error, I will try to fix it ASAP.

Done.

**Comment by Impyy** [comment]
Just tested this and it seems to work. 

It would be nice if this UPnP functionality could also map ``options.tcp_port`` to make it easier for users to host TCP relays without changing settings in their router.

**Comment by Ansa89** [comment]
> It would be nice if this UPnP functionality could also map `options.tcp_port` to make it easier for users
> to host TCP relays without changing settings in their router.

Partially done: everything is in the right place, but I cannot instruct the configure/build system (autoreconf?) to correctly build it.
You can find the new version [here](https://github.com/Ansa89/toxcore/tree/nat_test).

**Comment by Ansa89** [comment]
> It would be nice if this UPnP functionality could also map `options.tcp_port` to make it easier for users to host TCP relays without changing settings in their router.

Done.

File: toxcore/nat_traversal.c [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    char lanaddr[64];
+
+    error = UPNP_GetValidIGD(devlist, &urls, &data, lanaddr, sizeof(lanaddr));
+    freeUPNPDevlist(devlist);
+    if (error) {
+        if (error == 1) {
+            LOGGER_INFO("A valid IGD has been found.");
+
+            char portstr[10];
+            snprintf(portstr, sizeof(portstr), "%d", port);
```
Comment (cebe) [inline toxcore/nat_traversal.c]: I'm getting the following warning on make:

```
  CC       ../toxcore/libtoxcore_la-nat_traversal.lo
../toxcore/nat_traversal.c: In function 'upnp_map_port':
../toxcore/nat_traversal.c:70:13: warning: incompatible implicit declaration of built-in function 'snprintf'
             snprintf(portstr, sizeof(portstr), "%d", port);
             ^
```

**Comment by cebe** [comment]
Compiled this with uTox and UPNP works as expected. Code looks also good to me. My router does not support NAT-PMP so I could not test that.

**Comment by Ansa89** [comment]
@cebe: should be fixed now. Please test again.

**Comment by GrayHatter** [comment]
I'll test this some time this week!

**Comment by Impyy** [comment]
@GrayHatter any news?

**Comment by GrayHatter** [comment]
No, haven't had the time. Don't let that discourage you from bumping this
PR.

**Comment by GrayHatter** [comment]
Is there a way to disable UPnP at runtime?

Eg if I'm on a network I don't want to open ports on?

**Comment by cebe** [comment]
@GrayHatter set `traversal_type` option, when creating a new tox object: https://github.com/irungentoo/toxcore/pull/1539/files#diff-6ed9b859959c69656a6ec2bf5ba110adR421

**Comment by Ansa89** [comment]
What @cebe said is correct, but the correct link to the code is https://github.com/irungentoo/toxcore/pull/1539/files#diff-a58028aec4211a18b170732935eba2e3R445 (`tox.in.h` vs `tox.h`).

**Comment by GrayHatter** [comment]
this is now merged into `new`

Thanks @Ansa89

**Comment by Ansa89** [comment]
Nice, thanks.

**Pull request closed without merging**